smac_planner/src/a_star.cpp

```diff
@@ -386,7 +386,7 @@ AStarAlgorithm<NodeSE2>::NodePtr AStarAlgorithm<NodeSE2>::getAnalyticPath(
       Coordinates initial_node_coords = next->pose;
       proposed_coordinates = {static_cast<float>(reals[0]), static_cast<float>(reals[1]), angle};
       next->setPose(proposed_coordinates);
-      if (next->isNodeValid(_traverse_unknown, _collision_checker) && next != prev) {
+      if (next->isNodeValid(_traverse_unknown, _collision_checker) && next != prev && !next->wasVisited()) {
```

Comment:
This could create gaps in the path. Is there a way we want to handle that or just accept that the analytic expansions may have gaps?

I wish we could distinguish between nodes that were _visited_ and nodes that were in the _actual path_ that we were returning. We could have visited nodes that were part of dead expansions that would be OK to override and have no issues with looping.

Reply:
Another interesting point could be that we don't actually update the parents in this part of the loop. We don't need to check if it was visited here, this is just for checking if its valid. In L412, that's where we actually set the parents that could cause loops. By the time we're at L412, we know that the path is valid and we're going to return a analytic expansion that ends the search, so we could at that point cross reference if its visited, if its actually in the path we're returning to skip or not.

Reply:
I agree (esp with the bit about visited vs in the actual path). This code won't create gaps in the analytic expansion - it will abort the analytic expansion entirely if we come across a node that we already visited. L393-400 reset the coordinates then abort by returning nullptr.

The idea was that if we find a node that we have visited along the analytic expansion then it must be closer to the goal than the node we started the expansion from. But for some reason the search algorithm did not start the expansion there, so we respect that decision and abort completely.

Reply:
> it will abort the analytic expansion entirely if we come across a node that we already visited

I don't think we should do that. I think a node being visited shouldn't be of particular interest for the analytic expansions as long as all of them are valid.  I think the fact that it came across a visited node just means it needs to be careful how it plays with the parent to not create looping. 

I'd think the easiest solution would be to check if valid, but if visited, not append it (or more completely: if visited and not in path if rest of expansion is OK) to avoid looping but it still represents a valid path to goal to finish searching

Reply:
Latest commit moves the check for visitation, and skips the node if it was visited already. I will look into the full backtrace-and-check method next.